Fix various issues:

tractorcow opened the pull request:
- Use DI for creation to allow user-configuration of backends
- Correct some method signatures
- Fix bootstrapping issues with cookie key configuration
- Fix issues with database being operated on when not ready (such as when using testsession)
- Fix issue with cookie expiry of 0 being misinterpreted

Commit:
BUG Fix lifetime of DB store and ANSI SQL compatible syntax
BUG Fix config/db/flush bootstrapping issues
API Use DI for creation to allow user-configuration of backends
Fix some method signatures
BUG Fix issues with database being operated on when not ready (such as when using testsession)
BUG Fix issue with cookie expiry of 0 being misinterpreted
Add tests
Documentation

tractorcow commented:
I've moved the failure condition for missing cookie key to the constructor, which will prevent errors being initiated late in the request pipeline and blocking redirection of 'flush' parameter tokens.

tractorcow commented:
My review is pretty much complete. There are a lot of other things that I have changed, so come and chat to me if you have any concerns.

I've also added docs, composer.json, travis config, and tested it on my fork. See https://travis-ci.org/tractorcow/silverstripe-hybridsessions

tractorcow commented:
Now with more E_USER_WARNINGS

hafriedlander commented on code/HybridSessionStore.php:
```diff
@@ -178,78 +202,123 @@ public function decrypt($data) {
+			$expiry = $this->getNow() + $lifetime;
 
-			$cookie = $this->crypto->encrypt(
+			$this->currentCookieData = $this->crypto->encrypt(
```
Is this because you might get a read after a write? If so, $this->currentCookieData should be set to null if strlen($session_data) > 1024

hafriedlander commented on code/HybridSessionStore.php:
```diff
@@ -178,78 +202,123 @@ public function decrypt($data) {
 			);
 
+			// Skip write if too late to send cookie
+			if(headers_sent()) return false;
```
I don't understand why this was moved here - we end up creating an (expensive) crypto instance, just to set $this->currentCookieData, but I can't see any value to do that if the cookie isn't eventually set.

tractorcow commented:
I've updated and added tests to ensure that the cookie backend automatically invalidates any internally decrypted data that hasn't been written to the cookies.

hafriedlander commented on code/HybridSessionStore.php:
```diff
@@ -178,170 +202,277 @@ public function decrypt($data) {
+	public function read($session_id) {
+		// Check ability to safely decrypt content
+		$crypto = $this->getCrypto($session_id);
+		if(!$crypto || !$this->currentCookieData) return;
```
Crypto is expensive to construct - we shouldn't be creating it if we don't need it.

hafriedlander commented on code/HybridSessionStore.php:
```diff
@@ -178,170 +202,277 @@ public function decrypt($data) {
+		// Working around config handling of booleans
+		$override = Config::inst()->get(__CLASS__, 'override_headers_sent');
+		if(in_array($override, array('true', 'false'))) return $override === 'false';
+		return !headers_sent();
```
Using config like this seems very slow compared to just overwriting canWrite in a subclass when you want it to not call !headers_sent()
